### PR TITLE
feat: add --ns flag to override system (or google) resolvers

### PR DESCRIPTION
**Pull request by bungle**
This pull request adds a support for `--ns IP` flag to override `/etc/resolv.conf` or Google's fallback name servers. Why would we want this?

1. on development, you may want to use different resolver than what's found in `/etc/resolv.conf`
2. these days `/etc/resolv.conf` is not an authoritative source for name servers in many cases
3. you may want to use different name servers on different interfaces, different domains etc. (e.g. on macOS you do have `/etc/resolver/`directory where you can specify different name servers for say different top level domains.

This gives `resty` cli a little bit of flexibility.

### CODE REVIEW

File: bin/resty [hunk 1]
```diff
@@ -475,6 +481,8 @@ Options:
     --main-include path Include the specified file in the nginx main configuration block
                         (multiple instances are supported).
 
+    --resolver ip       Specify the name resolver(s) (multiple instances are supported)
```
Comment (agentzh) [inline bin/resty]: Shall we use all upper case for the placeholder value `ip`? That is, `--resolver IP`.

I know we also need to fix `--main-include path` above. But let's do it in a separate PR :)

**Comment by agentzh** [comment]
@bungle Looking good to me except that minor thing. Thanks for this. I was thinking about adding this feature as well ;)

@thibaultcha Mind to have a look at this PR? Thanks!

**Comment by thibaultcha** [comment]
LGTM! Just a couple of thoughts: should we keep the name `--resolver` or use `--nameserver`, and document it appropriately in the `--help` section as well: `Specify a name server (multiple instances are supported)`?

As the Nginx `resolver` directive is more implying that a DNS resolver will be created, with its own name servers and options, while with this option we are concerned about specifying name servers only. 

Just a thought, maybe we don't care about the terminology that much.

Another thought: other options accepting string values go through some sort of validation: `--shdict` is validated by a regex with `check_option_format()` before being included in the Nginx configuration, and `resolve_includes()` ensures the given files indeed exist. Should we validate the given IP addresses?

**Comment by agentzh** [comment]
@thibaultcha Good points! We should validate the option value format early for more user friendly error message. And yeah, we should make it explicit that we support multiple occurrences of this option. I'm not particularly worried about the option name though.

**Comment by bungle** [comment]
Is dependency to say `Net::IP` or `Data::Validate::IP` alright or should I just use regex? In my opinion lets keep `resolver` as it rims better with that `resolve-ipv6`.

**Comment by bungle** [comment]
@thibaultcha I did write a small test for resolver input parameters. I don't think we need more extensive validation here as this is not a default option, and people who use it should know what to input. IPv4 / IPv6 addresses are a horror to validate without extra dependencies.

**Comment by agentzh** [comment]
@bungle Some comments:

1. No, the `resty` tool should not depend anything other than the standard perl installation. Being as much portable as possible is a design goal.
2. The validation does not need to be thorough. For one thing, we want to avoid the user to inject arbitrary `nginx.conf` snippet by abusing this option. So a simple regex like `/^ (?: \d{1,3} (?: \. \d{1,3} ){3} | \[ [0-9a-fA-F:]+ \] )$/x`

**Comment by agentzh** [comment]
@bungle Also, I think maybe `--ns IP` is a better option name here?

**Comment by bungle** [comment]
@agentzh, changed made, should I also bump the version number, or will you do it?

**Comment by agentzh** [comment]
@bungle Don't worry about the version number. I'll take care of it :) Thanks!

File: bin/resty [hunk 2]
```diff
@@ -62,11 +62,13 @@ GetOptions("c=i",             \(my $conns_num),
            "nginx=s",         \(my $nginx_path),
            "valgrind",        \(my $use_valgrind),
            "valgrind-opts=s", \(my $valgrind_opts),
+           "ns=s",            \@ns,
            "resolve-ipv6",    \(my $resolve_ipv6),
            "V|v",             \(my $version))
    or usage(1);
 
 check_option_format('--shdict', qr/^[_a-z]+\s+\d+(?i)[km]?\z/, 'NAME SIZE', \@shdicts);
+check_option_format('--ns',     qr/^(?:\d{1,3}(?:\.\d{1,3}){3}|\[[0-9a-fA-F:]+\])$/x, 'IP', \@ns);
```
Comment (agentzh) [inline bin/resty]: Will you add spaces to the regex (as I did in one of my previous comments) to make it more readable? We are using `/x` here already, which means spaces in the regex are not significant (except those inside a character class). Also this line seems too long, better break it up.

File: t/resty/options.t [hunk 3]
```diff
@@ -324,3 +326,58 @@ print(str)
...
 resolver [\s\S]* ipv6=off;
 --- err
+
+
+
+=== TEST 23: specified nameserver overrides system nameservers
+--- opts: --ns 127.0.0.1
+--- src
+local prefix = ngx.config.prefix()
+local conf = prefix.."conf/nginx.conf"
```
Comment (agentzh) [inline t/resty/options.t]: Please use spaces around binary operators like `..`. Thank you. Please fix other similar places as well.

File: t/resty/options.t [hunk 4]
```diff
@@ -324,3 +326,58 @@ print(str)
...
+local f = assert(io.open(conf, "r"))
+local str = f:read("*a")
+f:close()
+print(str)
+--- out_like
+resolver 127.0.0.1 127.0.0.2 ipv6=off;
+--- err
+
```
Comment (agentzh) [inline t/resty/options.t]: Better test IPv6 addresses as well, for completeness.

File: bin/resty (hunk 2 above)
Comment (thibaultcha) [inline bin/resty]: +1 for this regex check. It's complete enough to prevent most user errors and makes use of already existing code instead of requiring an external dependencies :+1:

File: bin/resty [hunk 5]
```diff
@@ -475,6 +482,8 @@ Options:
     --main-include path Include the specified file in the nginx main configuration block
                         (multiple instances are supported).
 
+    --ns IP             Specify the nameserver(s) (multiple instances are supported)
```
Comment (thibaultcha) [inline bin/resty]: Shouldn't this be:
```
Specify a custom name server (multiple instances are supported)
```

- added "custom" to imply that by setting this field, the ones from `resolv.conf` won't be considered
- switched to singular, because this option only takes 1 ns, but can indeed be repeated
- "name server" instead of "nameserver" because it's the commonly used wording (including in the Nginx documentation of the `resolver` directive.

Just a thought, I don't see this as a blocker, just maybe an improvement :)

File: t/resty/options.t (hunk 3 above)
Comment (thibaultcha) [inline t/resty/options.t]: That's my bad, I believe @bungle took example on the previous tests up there ^ that I wrote for the `--resolve-ipv6` flag, and I didn't include spaces around `..`...

File: bin/resty (hunk 5 above)
Comment (bungle) [inline bin/resty]: My original regex supported multiple with one flag, but I will change this. I debated name server vs nameserver and the former feels generic, but the later is the thing (a technical entity), but whatever I can change them back.

File: bin/resty (hunk 2 above)
Comment (bungle) [inline bin/resty]: Are we strict here about 80 chars lines, well there are already longer ones.

**Comment by bungle** [comment]
We probably also want to fix the reading of `/etc/resolv.conf` as it doesn't currently support IPv6 name servers. Also our `regex` doesn't support `port` which may or may not be a concern. Then we might need to also match IPv6 with and without `[` and `]`, but add them for Nginx.

File: bin/resty (hunk 5 above)
Comment (bungle) [inline bin/resty]: Text changed to the one that @thibaultcha suggested.

**Comment by agentzh** [comment]
@bungle Merged. Thanks! Sorry for the delay on my side.
